sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala

```diff
@@ -83,6 +116,20 @@ object SubqueryExpression {
       case _ => false
     }.isDefined
   }
+
+  /**
+   * Clean the outer references by normalizing them to BindReference in the same way
+   * we clean up the arguments during LogicalPlan.sameResult. This enables to compare two
+   * plans which has subquery expressions. This method returns a [[CanonicalizedSubqueryExpr]]
+   * which wraps the underlying [[SubqueryExpression]].
+   */
+  def canonicalize(e: SubqueryExpression, attrs: AttributeSeq): CanonicalizedSubqueryExpr = {
+    // Normalize the outer references in the subquery plan.
+    val subPlan = e.plan.transformAllExpressions {
+      case OuterReference(r) => BindReferences.bindReference(r, attrs, allowFailures = true)
```

Comment:
@cloud-fan Hi Wenchen, I tried to normalize all the expressions in the plan. But still the hashcodes are not stable. On two instances of the same plan i get two different hashcode. Please advice.